### ISSUE TITLE
Problems overriding the default roles

### ISSUE THREAD

**Issue by akhmerov** [jupyterhub/jupyterhub#3703]
### Bug description
It seems that overriding the role `user` is not functioning correctly. Specifically, on jh 2.0.0 if I set up the following role config (see [here](https://discourse.jupyter.org/t/plans-on-bringing-rtc-to-jupyterhub/9813/11?u=anton_akhmerov) for motivation)

```python
c.JupyterHub.load_roles = [
    {
        "name": "user",
        "description": "Allow users to access the shared server in addition to default perms",
        "scopes": ["self", "access:servers!user=shared"],
    }
]
```

Now after starting the hub, any non-admin user that existed on the hub before the start is unable to launch their server and sees a 500 error:

![image](https://user-images.githubusercontent.com/2069677/144727721-6f6b5578-41f6-4704-9ffa-11c08428f7fc.png)

After deleting and recreating the user from the hub admin panel, everything seems to function normally.

**Comment by minrk** [comment]
@akhmerov thanks for the report! I'll dig into this.

In the meantime, do you have the log line that starts with `Token requesting role`? It should show specifically what permissions are conflicting.

**Comment by akhmerov** [comment]
Sure:
```
[W 2021-12-06 13:19:17.896 JupyterHub roles:436] Token requesting role server with scopes not held by owner shared: {'access:servers', 'users:activity', 'read:users:activity'}
```

**Comment by minrk** [comment]
Have you specified the `server` role as well?

**Comment by akhmerov** [comment]
No, what I have is the complete role spec. Is it necessary to set something for the server? I thought `server` inherits the subset of scopes. Plus if it was necessary to specify the server then it wouldn't explain why deleting and readding the user changes the behavior.

**Comment by minrk** [comment]
>  Is it necessary to set something for the server? 

You shouldn't need to, I'm just trying to track down the source of the problem, since I haven't been able to reproduce it myself just yet.

The relevant details:

1. `user` role defines the permissions held by all JupyterHub users
2. `server` role defines the permissions of the $JUPYTERHUB_API_TOKEN that the server itself uses. The [default][] is
    ```python
    [
                'users:activity!user',
                'access:servers!user',
    ],
    ```
   where `!user` is an alias for `!user=$name` for the _owner_ of the server

3. when issuing a token owned by a user, JupyterHub [checks][] that the token isn't being given permissions that its owner doesn't have.

What you are experiencing is the token issued during server startup for hte user `shared` is requesting the role `server, but the user `shared` doesn't have a superset of the the requested permissions for starting a server.

Here's a question: is the `shared` user in your `allowed_users` list? I believe that would cause the issue, because then the shared user might not have any role assigned. How are you starting the shared server?

[default]: https://github.com/jupyterhub/jupyterhub/blob/2.0.0/jupyterhub/roles.py#L50-L54
[checks]: https://github.com/jupyterhub/jupyterhub/blob/2.0.0/jupyterhub/roles.py#L402-L439

**Comment by akhmerov** [comment]
The problem doesn't only apply to the `shared` user. Rather all users except for admins encounter this problem with their own servers. It appears both if I attempt to launch a user's server via the admin panel or if a user tries to do it on their own.

Hmm, we don't have `allowed_users` explicitly specified in the config. Most users are created by the oauthenticator, and `shared` was created by the admin interface.

**Comment by minrk** [comment]
Ah, gotcha. and the issue mainly occurs on Hub restart? Then I think I know where to look.

My guess is that users are being stripped of their membership in the `user` role because you have defined a custom user role, but not specified any members. This would be the right thing to do if you had specified an explicit empty list (`"users": []`), but perhaps it behaves the same if the membership is simply omitted.

There are too many places user creation can happen!

**Comment by akhmerov** [comment]
Indeed, the problem occurs after hub restart with every non-admin user that is in the database before the hub has restarted. I have no custom roles defined, only `user`.

**Comment by minrk** [comment]
Yup, tracked it down - when you're using the "allow any authenticated user" mode, in combination with a custom user role, all users get their membership in `users` revoked at startup. Working on a fix.

No obvious workaround I can see, other than to specify a `c.Authenticator.allowed_users` set, if that's feasible.

**Comment by akhmerov** [comment]
Not restarting the hub in the is :ok_hand: as a workaround :laughing:

**Comment by minrk** [comment]
I believe this is fixed by #3708. I'll add a test to confirm before it's ready to merge.
